internal/app/go-filecoin/retrieval_market_connector/client.go

```diff
@@ -1,27 +1,292 @@
+			true,                                   // broadcast to network
+			paymentbroker.CreateChannel,            // command
+			fcMiner, validAt, // params: payment address, valid block height
+		)
+		r.mw.Wait(ctx, msgCid, r.handleMessageReceived)
+
+		return address.Undef, err
+	}
+
+	// Not a real actor, just plays one on PaymentChannels.
+	return chid, nil
+}
+
+// AllocateLane creates a new lane for this paymentChannel with 0 FIL in the lane
+// Assumes AllocateLane is called after GetOrCreatePaymentChannel
 func (r *RetrievalClientNodeConnector) AllocateLane(paymentChannel address.Address) (uint64, error) {
-	panic("TODO: go-fil-markets integration")
+	payer, err := r.getPayerFromChannelID(paymentChannel)
+	if err != nil {
+		return 0, err
+	}
+
+	if r.laneStore[payer] == nil {
+		r.laneStore[payer] = []types.AttoFIL{}
+	}
+	r.laneStore[payer] = append(r.laneStore[payer], types.NewAttoFILFromFIL(0))
+	lane := uint64(len(r.laneStore[payer]) - 1)
+
+	return lane, nil
+}
+
+// CreatePaymentVoucher tells the actor to create and sign a voucher and post this to the chain, returning
+// the signed voucher and any error.
+// If there is not enough value stored in the payment channel registry, an error is returned.
+// If a lane has not been allocated for this payment channel, an error is returned.
+func (r *RetrievalClientNodeConnector) CreatePaymentVoucher(ctx context.Context, paymentChannel address.Address, amount tokenamount.TokenAmount, lane uint64) (*gfm_types.SignedVoucher, error) {
+	payer, err := r.getPayerFromChannelID(paymentChannel)
+	if err != nil {
+		return nil, err
+	}
+
+	_, pinfo, err := r.pbAPI.GetPaymentChannelInfo(ctx, paymentChannel)
+	payChBal := pinfo.Amount.Sub(pinfo.AmountRedeemed)
+
+	if err != nil {
+		return nil, err
+	}
+	if payChBal.LessThan(types.NewAttoFIL(amount.Int)) {
+		return nil, errors.New("not enough funds in payment channel")
+	}
+
+	lanes, ok := r.laneStore[payer]
+	if !ok || len(lanes) == 0 {
+		return nil, errors.New("payment channel has no lanes allocated")
+	}
+
+	// set the allocated amount of the lane
+	lanes[lane] = types.NewAttoFIL(amount.Int)
+
+	v, err := r.createSignedVoucher(ctx, payer, lane, amount)
+	if err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+func (r *RetrievalClientNodeConnector) createSignedVoucher(ctx context.Context, payer address.Address, lane uint64, amount tokenamount.TokenAmount) (*gfm_types.SignedVoucher, error) {
+	height, err := r.getBlockHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	nonce, err := r.actAPI.NextNonce(ctx, payer)
+	if err != nil {
+		return nil, err
+	}
+
+	v := gfm_types.SignedVoucher{
+		TimeLock:       0,   // TODO
+		SecretPreimage: nil, // TODO
+		Extra:          nil, // TODO
+		Lane:           lane,
+		Nonce:          nonce,
+		Amount:         amount,
+		MinCloseHeight: height + 1,
+	}
+
+	var buf bytes.Buffer
+	if err := v.MarshalCBOR(&buf); err != nil {
+		return nil, err
+	}
+
+	signingAddr, err := r.wal.GetDefaultWalletAddress()
```

Comment:
I believe this voucher needs to be signed by the payer which is not necessarily the default wallet address.